llvm/test/CodeGen/X86/bfloat-calling-conv-no-sse2.ll

```diff
@@ -0,0 +1,1441 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=i386-linux-gnu < %s | FileCheck -check-prefixes=CHECK,NOSSE %s
+; RUN: llc -mtriple=i386-linux-gnu -mattr=+sse < %s | FileCheck -check-prefixes=CHECK,SSE %s
+
+; Make sure no assert without SSE2 and bfloat. Issue 92899
+
+define bfloat @return_arg_bf16(bfloat %x) {
+; CHECK-LABEL: return_arg_bf16:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    flds {{[0-9]+}}(%esp)
```

Comment:
This doesn't match with half, maybe fine if you just want to make assert happy.

Reply:
That's how it works in most target but doesn't work for X86, because when on 32-bit system or 64-bit without SSE, X87 register is used for `float` type. But X87 instruction cannot handle `half` type.

Reply:
But you could just as well replace half in ABI contexts with float, as is done for bfloat. You can make either decision, the instruction support doesn't really change this

Reply:
You are right, I mistook with another issue. You need to add back the code I removed in #76901. Even though, you probably still getting error on 64-bit like this https://godbolt.org/z/b36dxdsW9

Reply:
With this patch, I see that error. Without it, I see the assert this avoids. If you change your godbolt link to one of the llc builds with assertions, you also see the assert. 

Which code in #7609 do you mean? I'm not sure the removed splitValueIntoRegisterParts code matters here.

Reply:
I mean you need to add back splitValueIntoRegisterParts/joinRegisterPartsIntoValue, otherwise you just load/store in 32-bits width.

The assertions is not the key. The key is LLVM17 contains https://github.com/llvm/llvm-project/pull/76901 (removed since 18) which equals the patch is this case. What I want to show is even with this, we have another bar there.

Think again, I think this can be taken as a precedent how we handle ABI unsuppoted cases: https://godbolt.org/z/z5YdGdaWs
Maybe we just need to emit an error like that for BF16 without SSE2.

Reply:
Yes, the 32-bit load/store here is of a float value. It is not trying to load a bfloat as a 32-bit operation. The bfloat simply does not exist